addon/services/file-manager.js

```diff
@@ -1,165 +1,318 @@
     },
 
     checkout(/*file, user*/) {
         // TODO? Having checkout here makes more sense to me than making it
         // the only writable attribute on the file model.
     },
 
-    ///////////////////// Folder-only actions /////////////////////
-
-    addSubfolder(folder, name) {
+    /**
+     * Create a new folder
+     *
+     * @method addSubfolder
+     * @param {file} folder Location of the new folder, a `file` model with
+     * `isFolder == true`.
+     * @param {String} name Name of the folder to create.
+     * @param {Object} [options] Options hash
+     * @param {Object} [options.query] Key-value hash of query parameters to
+     * add to the request URL.
+     * @param {Object} [options.data] Payload to be sent.
+     * @return {Promise} Promise that resolves to the new folder's model or
+     * rejects with an error message.
+     */
+    addSubfolder(folder, name, options = {}) {
         let url = folder.get('links').new_folder;
-        let params = {
-            name,
-            kind: 'folder'
-        };
+        if (!options.query) {
+            options.query = {};
+        }
+        options.query.name = name;
+        options.query.kind = 'folder';
 
         // HACK: This is the only WB link that already has a query string
         let queryStart = url.search(/\?kind=folder$/);
         if (queryStart > -1) {
             url = url.slice(0, queryStart);
         }
-        return this._waterbutlerRequest('PUT', url, params);
+        let p = this._waterbutlerRequest('PUT', url, options);
+        return p.then(() => this._getNewFileInfo(folder, name));
     },
 
-    uploadFile(folder, name, contents) {
+    /**
+     * Upload a file
+     *
+     * @method uploadFile
+     * @param {file} folder Location of the new file, a `file` model with
+     * `isFolder == true`.
+     * @param {String} name Name of the new file.
+     * @param {Object} contents A native `File` object or another appropriate
+     * payload for uploading.
+     * @param {Object} [options] Options hash
+     * @param {Object} [options.query] Key-value hash of query parameters to
+     * add to the request URL.
+     * @param {Object} [options.data] Payload to be sent.
+     * @return {Promise} Promise that resolves to the new file's model or
+     * rejects with an error message.
+     */
+    uploadFile(folder, name, contents, options = {}) {
         let url = folder.get('links').upload;
-        let params = {
-            name,
-            kind: 'file'
-        };
-        return this._waterbutlerRequest('PUT', url, params, contents);
-    },
+        options.data = contents;
+        if (!options.query) {
+            options.query = {};
+        }
+        options.query.name = name;
+        options.query.kind = 'file';
 
-    ///////////////////// File and folder actions /////////////////////
+        let p = this._waterbutlerRequest('PUT', url, options);
+        return p.then(() => this._getNewFileInfo(folder, name));
```

Comment:
We talked about this in person, but for anyone else: The waterbutler entity returned for newly created files doesn't include the file ID used by the OSF API. In order to use the new file in any meaningful way, we need to get its metadata from the API by listing the contents of its parent folder.

For other (non-create) operations, we might technically have enough information to manually update our models to what they should be, but I'd still prefer reloading the model through the API, to be sure our models reflect what actually is.